torch/csrc/jit/codegen/cuda/transform_view.cpp

```diff
@@ -0,0 +1,682 @@
+#include <torch/csrc/jit/codegen/cuda/transform_view.h>
+
+#include <torch/csrc/jit/codegen/cuda/arith.h>
+#include <torch/csrc/jit/codegen/cuda/fusion.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_internal_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+#include <torch/csrc/jit/codegen/cuda/transform_iter.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+struct ViewIndexState {
+  // The index into the new view
+  size_t new_view_index = 0;
+
+  // The index into the original view
+  size_t original_view_index = 0;
+
+  // The index into the transform view
+  size_t transform_view_index = 0;
+
+  // The number of broadcast axes before this transformation
+  size_t broadcast_offset = 0;
+
+  // The number of trivial reduction axes before this transformation
+  size_t trivial_reduction_offset = 0;
+
+  // The number of split transformations
+  size_t split_offset = 0;
+
+  // The number of merge transformations
+  size_t merge_offset = 0;
+};
+
+//! Base class for all tranformations
+class Transform {
+ public:
+  virtual void toString(std::stringstream& output) const = 0;
+
+  size_t index() const {
+    return index_;
+  }
+
+ protected:
+  Transform(size_t index) : index_(index) {}
+  virtual size_t computeIndex(const ViewIndexState& state) = 0;
+  const size_t index_ = 0;
+};
+
+//! Base class for all view tranformations - Merge, Split, Keep
+//! These transforms require updating the rfactor domain of the view TensorView
+//! and are applied after removing any unnecessary trivial broadcasts.
+class ViewTransform : public Transform {
+ public:
+  virtual void createRfactorDomain(
+      const std::vector<IterDomain*>& new_root_domain,
+      std::vector<IterDomain*>& rfactor_domain) = 0;
+
+ protected:
+  ViewTransform(const ViewIndexState& state) : Transform(computeIndex(state)) {}
```

Comment:
This code pattern makes me feel rather uncomfortable as it calls the virtual function before initializing the parent class. I'm not sure this is guaranteed to work as intended.

Looks like it's not really necessary to make `computeIndex` a virtual function. I think it can be just a static function, and the pure virtual declaration in the parent class can be removed.

Reply:
FYI, why it's unsafe: https://stackoverflow.com/a/962148

Reply:
It is unsafe when calling the virtual function from inside the base class constructor. I called `computeIndex` inside the derived class constructor, so it will go to the derived class function.

I removed `computeIndex` the base class and made it a regular function in the derived classes.

Reply:
I don't think those exceptions apply here. The issue is the virtual function is used in the initialization list, so it is called before the parent class is constructed. In the exceptions listed in the page, they are used in the body of the constructor, so at that point the parent class is guaranteed to be constructed.